Removes incongruent priors from splicing output

falquaddoomi opened the pull request:
Adds blacklisting of problematic priors, specified via a new command line option, `--blacklisted_vars`. The value of the argument should be a file containing newline-delimited HGVS_cDNA specifiers for the variants. Variants with matching HGVS_cDNA column will return blankDict in `getVarData()`.

This PR also includes a list of blacklisted variants in `blacklisted_vars.txt`, extracted from a Google Doc that's been circulated internally. (Ask me if you'd like access, since the current permissions on it allow anyone to edit it.)

Note that I haven't tested this code; I was hoping someone who's able to run the pipeline more easily could give it a try.

rcurrie commented:
My preference is still to make this a separate step external to calcVarPriors, either in the pipeline itself or maybe better external. calVarPriors is complex enough as it is and needs to be re-factored. This addition required extending the partial() passed into the parallel mapping which itself is a hack vs. sorting out exactly what should be passed all over the place. That said I could live with doing it this way now in the spirit of getting this shipped and then we can refactor it along with the other big refactor once we sort what is .next for this pipeline. Defer to @zfisch and @falquaddoomi .

zfisch commented:
@falquaddoomi the most straightforward is probably to write a script that takes built_with_priors.tsv (as is in /home/pipeline/faisal) and blacklisted_vars.txt and wipes all priors data for blacklisted variants spitting out something like built_with_priors_clean.tsv (or some other such name). it's very straightforward to add this sort of script to the luigi pipeline (for clarification, the priors pipeline refers only to the code used to calculate priors, the luigi pipeline is the task runner that runs all steps including the priors pipeline and it's what we run monthly).

rcurrie commented:
That's sort of what I was thinking. Likely a few lines of pandas to read in the built_with_priors, redact based on the black list and emit a built_with_priors.redacted.tsv or equivalent...

falquaddoomi commented:
Agreed, having the filtering as a separate step is better. I originally implemented it in `calcVarPriors.py` because at the time I wasn't aware of the `integratePriorsIntoPipeline` branch which added the Luigi parts. Thankfully that branch has since been merged into master, so I rebased the changes in this PR off of it and added the relevant bits to Luigi. I named the post-filtering output `built_with_priors_clean.tsv` and updated the references to `built_with_priors.tsv` to that.

Some points:
- Since the filtering is its own step, it can't go inside the splicing Docker container (unless we modify the entrypoint, which I'm assuming we don't want to do). I left the script as well as its supporting files and tests in the splicing subdirectory since they're still logically part of that step.
- Pandas unfortunately reinterprets its input when converting a CSV to a DataFrame using pd.read_csv(). This causes some very minor rounding errors, which nevertheless cause it to fail the checksum and per-column matching tests. I've left the pandas implementation since it's about 7 seconds faster on my laptop (3sec for built_with_priors.tsv vs. 10 seconds for the csv implementation), but it's disabled by default. The unittest for the pandas implementation is also marked as "expected to fail"; if anyone manages to get it working in the future I suppose they can remove that and reuse the test to ensure it's working.
- It seems like there are two ways that testing is done in the splicing pipeline, via `unittest` and a 'test' argument to the script that validates a checksum of the output. I wasn't sure which was preferred, so I implemented both.

I haven't been able to run the full pipeline, so I'll probably need Marc or someone else to do so before I can't 100% guarantee that this is working, but the code is pretty simple and my own tests pass, so I'm fairly confident in it.

falquaddoomi commented on pipeline/splicingfilter/filterBlacklistedVars.py:
```diff
@@ -0,0 +1,197 @@
+
+@cli.command(help="Run self test")
+@click.pass_context
+def test(ctx):
```
I was mostly following the pattern I saw in the splicing code, which has both tests and a 'test' subcommand. I'm more than happy to remove it since the tests I currently have are, in my opinion, sufficient.

falquaddoomi commented on pipeline/splicingfilter/test_filterBlacklistedVars.py:
```diff
@@ -0,0 +1,41 @@
+
+class TestStringMethods(unittest.TestCase):
+    def filterShortPriors(self, method):
+        priorsPath = "tests/priors_short.tsv"
```
Ah, sorry about that; I think it's reasonable to use the path to the current module as you suggested.

falquaddoomi commented on pipeline/splicingfilter/filterBlacklistedVars.py:
```diff
@@ -0,0 +1,197 @@
+    return var_table.to_csv(path_or_buf=output, sep='\t', index=False, float_format='%s')
+
+
+def filter_vars_csv(variants, blacklisted_vars, output):
```
I wasn't sure if the loss of precision was acceptable, since the script using pandas would technically modify every field and not just the priors-relevant ones. I'll defer to your judgment and switch to pandas, though. Thanks for the code sample, too.
